docs: Upgrade docs for v1beta1 (#4749)
aws/karpenter: website/content/en/preview/upgrade/upgrade-ref.md

> ---

Author: Done.

> ---

Author: Okay. I also changed two links to this page from the guide.

title: "Karpenter Upgrade Reference"
linkTitle: "Karpenter Upgrade Reference"
weight: 10
description: >
  API information for upgrading Karpenter
---

> Significant changes to the Karpenter APIs and various compatibility issues have been introduced in Karpenter v0.32.

Reviewer: ```suggestion
Significant changes to the Karpenter APIs have been introduced in Karpenter v0.32.x.
```

Author: Done.

In this release, Karpenter APIs have advanced to v1beta1, in preparation for Karpenter v1 in the near future.
The v1beta1 changes are meant to simplify and improve ease of use of those APIs, as well as solidify the APIs for the v1 release.
Use this document as a reference to the changes that were introduced in the current release and as a guide to how you need to update the manifests and other Karpenter objects you created in previous Karpenter releases.

> The [Karpenter Upgrade Guide}(]({{< relref "upgrade-guide" >}}) steps you through the process of upgrading Karpenter for the latest release.

Reviewer: ```suggestion
The [Karpenter Upgrade Guide]({{< relref "upgrade-guide" >}}) steps you through the process of upgrading Karpenter for the latest release.
```

Author: Done.

> For a more general understanding of Karpenter compatibility issues, see the ({{< relref "#compatibility" >}}) section of this document.

Reviewer: ```suggestion
For a more general understanding of Karpenter's compatibility, see the [Compatibility Document]({{< relref "compatibility" >}}).
```

Author: Done.

# Karpenter Migration Information

Use the information below to help migrate your Karpenter v1alpha assets to v1beta1.

### Custom Resource Definition (CRD) Upgrades

> Karpenter ships with a few Custom Resource Definitions (CRDs). Starting with v0.32, CRDs representing Provisioners (karpenter.sh_provisioners.yaml), Machines (karpenter.sh_machines.yaml) and AWS Node Templates (karpenter.k8s.aws_awsnodetemplates.yaml) are replaced by those for Node Pools, Node Claims, and EC2 Node classes, respectively.

Reviewer: ```suggestion
Karpenter ships with a few Custom Resource Definitions (CRDs). Starting with v0.32.0, CRDs representing Provisioners, Machines, and AWS Node Templates are replaced by those for Node Pools, Node Claims, and EC2 Node classes, respectively.
```

Do you feel like the yaml link specified in parenthesis is neccessary?

Author: We can drop them. Done.

You can find these CRDs on [Karpenter APIs CRD](https://github.com/aws/karpenter/tree/main/pkg/apis/crds) page.

> The [Karpenter Upgrade Guide}(]({{< relref "upgrade-guide" >}}) describes how to install the new CRDs using `kubectl`. However, you can instead install those CRDs as an independent helm chart [karpenter-crd](https://gallery.ecr.aws/karpenter/karpenter-crd), that can be used by Helm to manage the lifecycle of these CRDs.  See [source](https://github.com/aws/karpenter/blob/main/charts/karpenter-crd) for the source code,

Reviewer: Should this be part of the upgrade guide? I recalled us originally talking about adding this to the FAQ

Author: I thought it went better here, since I don't think people would look for that information in the FAQ, but might run into it here as an alternative.

To upgrade or install `karpenter-crd` run:

```
helm upgrade --install karpenter-crd oci://public.ecr.aws/karpenter/karpenter-crd --version vx.y.z --namespace karpenter --create-namespace
```

{{% alert title="Note" color="warning" %}}
< If you get the error `invalid ownership metadata; label validation error:` while installing the `karpenter-crd` chart from an older version of Karpenter, follow the [Troubleshooting Guide]({{<ref "../troubleshooting#helm-error-when-installing-the-karpenter-crd-chart" >}}) for details on how to resolve these errors.
{{% /alert %}}

## Annotations, Labels, and Status Conditions

Karpenter v1beta1 introduces changes to some common labels, annotations, and status conditions that are present in the project. The tables below lists the v1alpha5 values and their v1beta1 equivalent.

> | Core Karpenter Labels           |                             |

Reviewer: ```suggestion
| Karpenter Labels           |                             |
```

Drop the "Core" prefix

Author: Done.

|---------------------------------|-----------------------------|
| **v1alpha5**                    | **v1beta1**                 |
| karpenter.sh/provisioner-name   | karpenter.sh/nodepool       |

> | Core Karpenter Annotations      |                             |

Reviewer: ```suggestion
| Karpenter Annotations      |                             |
```

Drop the "Core" prefix

Author: Done.

|---------------------------------|-----------------------------|
| **v1alpha5**                    | **v1beta1**                 |

> | karpenter.sh/provisioner-hash   | karpenter.sh/nodepool-hash  |

Reviewer: We need to add a line to this

```
| karpenter.k8s.aws/nodetemplate-hash | karpenter.k8s.aws/nodeclass-hash |
```

Author: Okay with me.

Author: Done.

  kubeletConfiguration:
    systemReserved:
      cpu: 100m
      memory: 100Mi
      ephemeral-storage: 1Gi
    maxPods: 20
  limits:
    resources:
      cpu: 1000
      memory: 1000Gi
  weight: 50
```

**NodePool example (v1beta1)**

```
apiVersion: karpenter.sh/v1beta1
kind: NodePool
...
spec:
  template:
    metadata:
      annotations:
        custom-annotation: custom-value
      labels:
        team: team-a
        custom-label: custom-value
    spec:
      requirements:
      - key: karpenter.k8s.aws/instance-generation
        operator: Gt
        values: ["3"]
      - key: karpenter.k8s.aws/instance-category
        operator: In
        values: ["c", "m", "r"]
      - key: karpenter.sh/capacity-type
        operator: In
        values: ["spot"]
      taints:
      - key: example.com/special-taint
        value: "true"
        effect: NoSchedule
      startupTaints:
      - key: example.com/another-taint
        value: "true"
        effect: NoSchedule
      kubeletConfiguration:
        systemReserved:
          cpu: 100m
          memory: 100Mi
          ephemeral-storage: 1Gi
        maxPods: 20
  limits:
    cpu: 1000
    memory: 1000Gi
  weight: 50

> ```

Reviewer: Do we want to provide any detail to describe the difference here. Something like, "Note that the `Limits` and `Weight` fields sit outside of the template section. The `Labels` and `Annotations` fields from the Provisioner are now under the `spec.template.metadata` section. Note that all other fields including requirements, taints, kubeletConfiguration, etc. are specified under the `spec.template.spec` section."

Author: Done.

### Provider

The Karpenter `spec.provider` field has been deprecated since version v0.7.0 and is now removed in the new `NodePool` resource. Any of the fields that you could specify within the `spec.provider` field are now available in the separate `NodeClass` resource.

**Provider example (v1alpha)**

```
apiVersion: karpenter.sh/v1alpha5
kind: Provisioner
...
spec:
  provider:
    amiFamily: Bottlerocket
    tags:
      test-tag: test-value  
```

**Nodepool example (v1beta1)**

```
apiVersion: karpenter.sh/v1beta1
kind: NodePool
...
nodeClassRef:
  name: default

> ```

Reviewer: Should we collapse these into a single section like

```yaml
apiVersion: karpenter.sh/v1beta1
kind: NodePool
...
nodeClassRef:
  name: default
---
apiVersion: karpenter.k8s.aws/v1beta1
kind: EC2NodeClass
metadata:
  name: default
spec:
  amiFamily: Bottlerocket
  tags:
    test-tag: test-value
```

Author: I like it as two section because it makes it clear that they are two separate things.

As part of the v1beta1 migration, Karpenter has chosen to collapse the concepts of emptiness and underutilization into a single concept: `consolidation`.
You can now define the types of consolidation that you want to support in your `consolidationPolicy` field.
The current values for this field are `WhenEmpty` or `WhenUnderutilized` (defaulting to `WhenUnderutilized` if not specified).
If specifying `WhenEmpty`, you can define how long you wish to wait for consolidation to act on your empty nodes by tuning the `consolidateAfter` parameter.
This field works the same as the `ttlSecondsAfterEmpty` field except this field now accepts either of the following values:

* `Never`: This disables empty consolidation.
* Duration String (e.g. “10m”, “1s”): This enables empty consolidation for the time specified.

**ttlSecondsAfterEmpty example (v1alpha)**

```
apiVersion: karpenter.sh/v1alpha5
kind: Provisioner
...
spec:
  ttlSecondsAfterEmpty: 120
```

**consolidationPolicy and consolidateAfter examples (v1beta1)**
```
apiVersion: karpenter.sh/v1beta1
kind: NodePool
...
spec:
  disruption:
    consolidationPolicy: WhenEmpty
    consolidateAfter: 2m

```

### Consolidation

> The Karpenter `spec.consolidation` block has been shifted into a `consolidationPolicy`. If you were previously enabling Karpenter’s consolidation feature for underutilized nodes using the `consolidation.enabled` flag, you now enable consolidation through the `consolidationPolicy`.

Reviewer: ```suggestion
The Karpenter `spec.consolidation` block has also been shifted under `consolidationPolicy`. If you were previously enabling Karpenter’s consolidation feature for underutilized nodes using the `consolidation.enabled` flag, you now enable consolidation through the `consolidationPolicy`.
```

Author: Done.

**consolidation enabled example (v1alpha)**

```
apiVersion: karpenter.sh/v1alpha5
kind: Provisioner
...
spec:
  consolidation:
    enabled: true
```

**consolidationPolicy WhenUnderutilized example (v1beta1)**

```
apiVersion: karpenter.sh/v1beta1
kind: NodePool
...
spec:
  disruption:
    consolidationPolicy: WhenUnderutilized

> 

Reviewer: cut newline

Author: Done.

```

> > Note: You currently can’t the `consolidateAfter` field when specifying `consolidationPolicy: WhenUnderutilized`. Karpenter will use a 15s `consolidateAfter` runtime default. 

Reviewer: ```suggestion
> Note: You currently can’t set the `consolidateAfter` field when specifying `consolidationPolicy: WhenUnderutilized`. Karpenter will use a 15s `consolidateAfter` runtime default. 
```

```
apiVersion: karpenter.k8s.aws/v1alpha1
kind: AWSNodeTemplate
...
spec:
  amiSelector:
    karpenter.sh/discovery: karpenter-demo
```

**amiSelectorTerms.tags example (v1beta1)**
```
apiVersion: compute.k8s.aws/v1beta1
kind: EC2NodeClass
...
spec:
  amiSelectorTerms:
  - tags:
      karpenter.sh/discovery: karpenter-demo
```

**amiSelector example (v1alpha)**
```
apiVersion: karpenter.k8s.aws/v1alpha1
kind: AWSNodeTemplate
...
spec:
  amiSelector:
    aws::ids: ami-123,ami-456
```

**amiSelectorTerms example (v1beta1)**
```
apiVersion: compute.k8s.aws/v1beta1
kind: EC2NodeClass
...
spec:
  amiSelectorTerms:
  - id: ami-123
  - id: ami-456
```

**amiSelector example (v1alpha)**
```
apiVersion: karpenter.k8s.aws/v1alpha1
kind: AWSNodeTemplate
...
spec:
  amiSelector:
    aws::name: my-name1,my-name2
    aws::owners: 123456789,amazon
```

**amiSelectorTerms.name example (v1beta1)**
```
apiVersion: compute.k8s.aws/v1beta1
kind: EC2NodeClass
...
spec:
  amiSelectorTerms:
  - name: my-name1
    owner: 123456789
  - name: my-name2
    owner: 123456789
  - name: my-name1
    owner: amazon
  - name: my-name2
    owner: amazon
```

### LaunchTemplateName

> The `spec.launchTemplateName` field for referencing unmanaged launch templates within Karpenter has been removed. 

Reviewer: Do we want to provide any more detail on this? Perhaps we could link to the design doc that proposes and confirms that we decided to drop this field.

Author: I added a link to the design doc: https://github.com/aws/karpenter/blob/main/designs/unmanaged-launch-template-removal.md

In addition to these metrics, the MachineNotFound error returned by the `karpenter_cloudprovider_errors_total` values in the error label has been changed to `NodeClaimNotFound`. This is agnostic to the version of the API (Machine or NodeClaim) that actually owns the instance.

## Global Settings

The v1beta1 specification removes the `karpenter-global-settings` ConfigMap in favor of setting all Karpenter configuration using environment variables. Along, with this change, Karpenter has chosen to remove certain global variables that can be configured with more specificity in the EC2NodeClass . These values are marked as removed below.

| **`karpenter-global-settings` ConfigMap Key**     | **Environment Variable**        | **CLI Argument**
|---------------------------------------------------|---------------------------------|-----------------------------------|
| batchMaxDuration                                  | BATCH_MAX_DURATION              | --batch-max-duration              |
| batchIdleDuration                                 | BATCH_IDLE_DURATION             | --batch-idle-duration             |

> | aws.assumeRoleARN                                 | AWS_ASSUME_ROLE_ARN             | --aws-assume-role-arn             |

Reviewer: We decided to remove the `aws` prefix from all of these CLI args and environment variables. Can you drop the aws prefix from all things in this table

Author: Sure. Done.

| aws.enablePodENI                                  | Dropped                         | Dropped                           |
| aws.enableENILimitedPodDensity                    | Dropped                         | Dropped                           |
| aws.isolatedVPC                                   | AWS_ISOLATED_VPC                |--aws-isolated-vpc                 |
| aws.vmMemoryOverheadPercent                       | AWS_VM_MEMORY_OVERHEAD_PERCENT  |--aws-vm-memory-overhead-percent   |
| aws.interruptionQueueName                         | AWS_INTERRUPTION_QUEUE_NAME     |--aws-interruption-queue-name      |
| featureGates.enableDrift                          | FEATURE_GATE="Drift=true"       |--feature-gates Drift=true         |

## Drift Enabled by Default

The drift feature will now be enabled by default starting from v0.33.0. If you don’t specify the Drift featureGate, the feature will be assumed to be enabled. You can disable the drift feature by specifying --feature-gates Drift=false. This feature gate is expected to be dropped when core APIs (NodePool, NodeClaim) are bumped to v1.

> 

Reviewer: I don't see mention of the updating sections that were added at the bottom of the draft migration proposal doc. There's a few other things at the bottom of that doc that we should add in here including:
1. AMI Requirements Tagging was dropped
2. Naming on Instances changed to nodeName
3. Logging Configuration is Deprecated
4. Logging Configuration is No Longer Dynamic
5. etc.
